Upgrade 3B — Extractor Reachability and Persistence Hygiene Patch

## Task class

source task / test task

## Mode

bounded implementation + verification

## Scope class

engineering-grade evidence hygiene / extractor reachability and persistence contract

## Objective

Close the low-risk, high-trust evidence hygiene gaps found by #105 before deeper extractor or Document Center upgrades.

This packet must fix or explicitly prove the current extractor reachability and persistence contract so SerapeumAI does not silently imply unsupported evidence capabilities.

## Current authority

```text
main includes PR #104 / #103 completion
#105 engineering-grade evidence baseline audit completed
standing backlog = #24
workspace honesty rail = complete through cross-project isolation proof
```

#105 audit verdict:

```text
Evidence surfaces exist, but extractor reachability, persistence completeness, metadata completeness, and dependency honesty remain publish-critical gaps.
```

## Why this is next

The File Inspector four-lane evidence surface exists and is tested. PDF and P6 evidence paths are partially real. But #105 found several hygiene gaps that are dangerous before deeper implementation:

1. Some extractor classes exist but are not reachable through `ExtractJob.EXTRACTORS`.
2. IFC extractor emits record types that `ExtractJob._insert_record(...)` does not persist.
3. Excel register extractor writes debug output to an absolute path:

```text
D:\SerapeumAI\extractor_debug.txt
```

These are not feature-expansion problems. They are evidence honesty/hygiene problems.

## Allowed scope

Minimal source/test changes only.

Likely touched files:

```text
src/application/jobs/extract_job.py
src/engine/extractors/register_extractor.py
src/tests/test_extractor_registry_reachability.py
src/tests/test_ifc_extractor_persistence_contract.py
src/tests/test_excel_register_extractor_hygiene.py
```

Allowed work:

- make extractor registry reachability explicit and tested;
- either register reachable extractors or explicitly test/document why an extractor is not mounted;
- remove absolute debug-file writing from Excel register extractor;
- ensure IFC emitted record types are either persisted or explicitly diagnosed rather than silently dropped;
- add focused tests for these contracts.

## Forbidden scope

Do not widen into evidence feature expansion.

Forbidden:

- no packaging changes;
- no dependency changes;
- no OCR dependency additions;
- no IFC/P6 parser dependency additions;
- no runtime/provider/provisioning work;
- no quantization/model-catalog work;
- no chat tool execution bridge;
- no LLM parser;
- no MCP/autonomous loop;
- no audit persistence implementation;
- no snapshot implementation;
- no broad UI redesign;
- no Document Center redesign;
- no Schedule Truth Workspace implementation;
- no Revit bridge work.

## Required behavior

### 1. Extractor registry reachability

`ExtractJob.EXTRACTORS` must have an explicit tested contract.

Current audit noted that these extractor classes exist:

```text
WordExtractor
PPTXExtractor
DGNExtractor
```

but the registry currently includes only:

```text
p6
ifc
excel_register
pdf
field
```

Required outcome:

- Either register Word/PPTX/DGN if they are intended as mounted extraction paths;
- or add a focused test/policy contract proving they are intentionally not mounted yet.

Preferred outcome:

```text
Register WordExtractor, PPTXExtractor, and DGNExtractor in ExtractJob.EXTRACTORS if imports are safe and no dependency expansion is needed.
```

### 2. Excel register debug-path hygiene

`ExcelRegisterExtractor` must not write to an absolute debug path.

Remove or replace:

```text
D:\SerapeumAI\extractor_debug.txt
```

Acceptable replacement:

- use `logger.debug(...)`;
- use diagnostics list;
- no filesystem debug side effect.

### 3. IFC emitted record persistence contract

`IFCExtractor` emits:

```text
ifc_project
ifc_spatial
ifc_element_metadata
ifc_connection
```

`ExtractJob._insert_record(...)` currently persists only part of that set.

Required outcome:

- Persist `ifc_element_metadata` and `ifc_connection` if schema supports it;
- or explicitly diagnose/guard unsupported IFC record types so they are not silently dropped.

Do not add schema unless absolutely necessary and explicitly justified. Prefer test-proven diagnostics if persistence tables do not exist.

### 4. No new dependencies

Do not install or require new libraries. If an extractor depends on optional libraries, this packet should make that dependency honesty clearer but must not add dependencies.

## Acceptance criteria

This issue passes only if:

1. ExtractJob extractor registry contract is tested.
2. Word/PPTX/DGN reachability is either added or explicitly not-mounted by test/policy.
3. Excel register extractor no longer writes to absolute debug file path.
4. IFC unsupported emitted record types are not silently dropped.
5. Existing evidence and workspace tests still pass.
6. No packaging/dependency/runtime files are touched.
7. Any source changes are minimal and directly tied to #105 hygiene findings.

## Required verification commands

Run from repo root on Windows PowerShell:

```powershell
python -m pytest -q src/tests/test_extractor_registry_reachability.py src/tests/test_ifc_extractor_persistence_contract.py src/tests/test_excel_register_extractor_hygiene.py src/tests/test_file_inspector_evidence_lanes.py src/tests/test_file_inspector_source.py src/tests/test_p6_truth.py src/tests/test_retrieval_project_isolation.py src/tests/test_mounted_chat_wiring.py src/tests/test_multi_lane_answer_path.py src/tests/test_truth_state_enforcement.py src/tests/test_snapshot_state_wording_honesty.py
```

If the final test names differ, include the equivalent focused tests.

Then inspect changed files:

```powershell
git diff --name-only
```

Expected files should stay inside the allowed source/test surfaces.

## Expected output from executor

Return:

- files touched;
- exact diff summary;
- exact test output;
- whether any extractor was registered or explicitly parked;
- whether IFC unsupported records are persisted or diagnosed;
- residual risks;
- PASS/FAIL verdict.

## Risks

- Packaging risk: low if no dependencies are added.
- Windows risk: low; removing absolute debug path reduces Windows portability risk.
- Rollback risk: medium if registry behavior changes.
- Architecture risk: reduced by making extractor reachability explicit.
- Product risk: reduced by avoiding silent evidence capability drift.

## Exit condition

Close only after extractor reachability, Excel debug-path hygiene, and IFC emitted-record handling are tested and proven.

## Follow-up after this issue

Next likely packet after closure:

```text
Upgrade 3C — PDF Metadata Completeness Contract
```

or a narrower packet selected by #3B findings.

## Local #106 verification failed after first branch push

Owner-run Windows PowerShell proof shows the branch was pushed, but the required verification did **not** pass.

Failing command:

```powershell
python -m pytest -q `
  src/tests/test_extractor_registry_reachability.py `
  src/tests/test_ifc_extractor_persistence_contract.py `
  src/tests/test_excel_register_extractor_hygiene.py `
  src/tests/test_file_inspector_evidence_lanes.py `
  src/tests/test_file_inspector_source.py `
  src/tests/test_p6_truth.py `
  src/tests/test_retrieval_project_isolation.py `
  src/tests/test_mounted_chat_wiring.py `
  src/tests/test_multi_lane_answer_path.py `
  src/tests/test_truth_state_enforcement.py `
  src/tests/test_snapshot_state_wording_honesty.py
```

Observed result:

```text
1 failed, 33 passed, 5 warnings in 30.22s
```

Failure:

```text
FAILED src/tests/test_ifc_extractor_persistence_contract.py::test_ifc_element_metadata_records_are_persisted_to_ifc_elements
sqlite3.IntegrityError: FOREIGN KEY constraint failed
```

Interpretation:

- Source patch intent remains valid.
- The test fixture is incomplete: it creates the baseline schema but does not seed the `file_registry` / `file_versions` row referenced by `ExtractJob.file_version_id = fv_ifc`.
- Do not open a PR until the branch is amended and the full focused verification passes.

Required fix:

- Update `src/tests/test_ifc_extractor_persistence_contract.py` fixture to seed `file_registry` and `file_versions` for `fv_ifc`.
- Re-run full #106 verification.
- Amend the existing commit and force-push the branch only after tests pass.

## Local #106 verification failed again after amend/force-push

Owner-run Windows PowerShell proof shows the amended branch was force-pushed, but required verification still did **not** pass.

Observed failure remains:

```text
FAILED src/tests/test_ifc_extractor_persistence_contract.py::test_ifc_element_metadata_records_are_persisted_to_ifc_elements
sqlite3.IntegrityError: FOREIGN KEY constraint failed

1 failed, 33 passed, 5 warnings
```

Important detail:

- The attempted fixture replacement reported:

```text
[FAIL] Expected fixture block not found. Showing file context:
Stop. Fixture block not found.
```

- The script then continued, but the replacement was effectively not applied.
- The branch was amended and force-pushed despite failing tests.

Current status:

```text
#106 is still open / not PR-ready.
Do not open PR until the full focused verification passes.
```

Required next action:

- Overwrite `src/tests/test_ifc_extractor_persistence_contract.py` with a corrected fixture that seeds `file_registry` and `file_versions`.
- Re-run full #106 verification.
- Amend and force-push only after tests pass.

--- a/src/application/jobs/extract_job.py
+++ b/src/application/jobs/extract_job.py
@@ -11,6 +11,9 @@ from src.engine.extractors.ifc_extractor import IFCExtractor
 from src.engine.extractors.register_extractor import ExcelRegisterExtractor
 from src.engine.extractors.pdf_extractor import UniversalPdfExtractor
 from src.engine.extractors.field_extractor import FieldExtractor
+from src.engine.extractors.word_extractor import WordExtractor
+from src.engine.extractors.pptx_extractor import PPTXExtractor
+from src.engine.extractors.dgn_extractor import DGNExtractor
 
 logger = logging.getLogger(__name__)
 
@@ -34,7 +37,10 @@ class ExtractJob(Job):
         "ifc": IFCExtractor,
         "excel_register": ExcelRegisterExtractor,
         "pdf": UniversalPdfExtractor,
-        "field": FieldExtractor
+        "field": FieldExtractor,
+        "word": WordExtractor,
+        "pptx": PPTXExtractor,
+        "dgn": DGNExtractor,
     }
 
     # ... (existing code for init/to_dict/from_dict) ...
@@ -245,6 +251,61 @@ class ExtractJob(Job):
              db.execute("INSERT OR REPLACE INTO ifc_projects (global_id, file_version_id, name) VALUES (?, ?, ?)", (data["GlobalId"], vid, data["Name"]))
         elif rtype == "ifc_spatial":
              db.execute("INSERT OR REPLACE INTO ifc_spatial_structure (element_id, file_version_id, entity_type, name) VALUES (?, ?, ?, ?)", (data["GlobalId"], vid, data["EntityType"], data["Name"]))
+        elif rtype == "ifc_element_metadata":
+             element_id = data.get("ElementId")
+             if not element_id:
+                 logger.warning("[ExtractJob] Skipping IFC element metadata without ElementId: %s", data)
+                 return
+             existing = db.execute(
+                 "SELECT raw_properties_json FROM ifc_elements WHERE element_id=? AND file_version_id=?",
+                 (element_id, vid),
+             ).fetchone()
+             raw_properties = {}
+             if existing and existing["raw_properties_json"]:
+                 try:
+                     raw_properties = json.loads(existing["raw_properties_json"]) or {}
+                 except Exception:
+                     raw_properties = {"_previous_raw": existing["raw_properties_json"]}
+             pset_name = data.get("PSetName") or "Properties"
+             raw_properties[pset_name] = {
+                 "properties": data.get("Properties", {}),
+                 "is_quantity": data.get("IsQuantity", False),
+             }
+             db.execute(
+                 "INSERT OR REPLACE INTO ifc_elements (element_id, file_version_id, spatial_container_id, entity_type, name, tag, raw_properties_json) VALUES (?, ?, ?, ?, ?, ?, ?)",
+                 (
+                     element_id,
+                     vid,
+                     data.get("SpatialContainerId"),
+                     data.get("EntityType"),
+                     data.get("Name") or data.get("PSetName"),
+                     data.get("Tag"),
+                     json.dumps(raw_properties),
+                 ),
+             )
+        elif rtype == "ifc_connection":
+             element1 = data.get("Element1Id")
+             element2 = data.get("Element2Id")
+             if not element1 or not element2:
+                 logger.warning("[ExtractJob] Skipping IFC connection without both endpoint ids: %s", data)
+                 return
+             link_id = f"ifc_conn_{vid}_{element1}_{element2}"
+             db.execute(
+                 "INSERT OR REPLACE INTO links (link_id, project_id, link_type, from_kind, from_id, to_kind, to_id, status, confidence, method_id, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
+                 (
+                     link_id,
+                     self.project_id,
+                     "ifc.connection",
+                     "ifc_element",
+                     element1,
+                     "ifc_element",
+                     element2,
+                     "CANDIDATE",
+                     1.0,
+                     "ifc_extractor_v1",
+                     db._ts(),
+                 ),
+             )
              
         # Register Logic
         elif rtype == "register_row":

--- a/src/engine/extractors/register_extractor.py
+++ b/src/engine/extractors/register_extractor.py
@@ -75,55 +75,57 @@ class ExcelRegisterExtractor(BaseExtractor):
         diagnostics = []
         
         
-        with open("D:\\SerapeumAI\\extractor_debug.txt", "a", encoding="utf-8") as dbg:
-            dbg.write(f"\n--- Extracting {file_path} ---\n")
-            try:
-                # Load all sheets
-                xls = pd.read_excel(file_path, sheet_name=None, dtype=str)
-                dbg.write(f"Sheets found: {list(xls.keys())}\n")
-                
-                for sheet_name in xls.keys():
-                    # Read first 30 rows to sniff header
-                    df_peek = pd.read_excel(file_path, sheet_name=sheet_name, nrows=30, header=None, dtype=str)
-                    header_row_idx = self._detect_header_row(df_peek)
-                    
-                    dbg.write(f"Sheet {sheet_name}: Detected Header at Row {header_row_idx}\n")
-                    
-                    # Read full sheet with correct header
-                    # If header_row_idx is None, assume 0 or empty sheet
-                    start_row = header_row_idx if header_row_idx is not None else 0
-                    df = pd.read_excel(file_path, sheet_name=sheet_name, header=start_row, dtype=str)
-                    
-                    if header_row_idx is None and len(df) < 2:
-                         dbg.write(f"Sheet {sheet_name}: Empty or unstructured, skipping.\n")
-                         continue
+        try:
+            # Load all sheets once, then inspect each sheet header before full extraction.
+            # Diagnostics are kept in-memory and through logger only; no debug file is written.
+            xls = pd.read_excel(file_path, sheet_name=None, dtype=str)
+            diagnostics.append(f"Sheets found: {list(xls.keys())}")
+            logger.debug("Excel register sheets found in %s: %s", file_path, list(xls.keys()))
 
-                    # Clean keys
-                    df.columns = [str(c).strip() for c in df.columns]
-                    
-                    # Iterate rows
-                    for idx, row in df.iterrows():
-                        # Remove NaNs / Nones
-                        row_data = {k: v for k, v in row.to_dict().items() if pd.notna(v) and str(v) != "nan" and v != ""}
-                        
-                        if not row_data:
-                            continue # Skip empty rows
-                            
-                        records.append({
-                            "type": "register_row",
-                            "data": {
-                                "sheet_name": sheet_name,
-                                "row_index": idx + start_row + 1, # Adjust for skip
-                                "content": row_data
-                            },
-                            "provenance": {"sheet": sheet_name, "row": idx + start_row + 1}
-                        })
-                dbg.write(f"Total records: {len(records)}\n")
-                return ExtractionResult(records=records, diagnostics=diagnostics, success=True)
-                
-            except Exception as e:
-                dbg.write(f"ERROR: {e}\n")
-                logger.error(f"Excel Extraction failed: {e}")
-                import traceback
-                dbg.write(traceback.format_exc())
-                return ExtractionResult(success=False, diagnostics=[str(e)])
+            for sheet_name in xls.keys():
+                # Read first 30 rows to sniff header.
+                df_peek = pd.read_excel(file_path, sheet_name=sheet_name, nrows=30, header=None, dtype=str)
+                header_row_idx = self._detect_header_row(df_peek)
+                diagnostics.append(f"Sheet {sheet_name}: detected header row {header_row_idx}")
+                logger.debug("Excel register sheet %s detected header row %s", sheet_name, header_row_idx)
+
+                start_row = header_row_idx if header_row_idx is not None else 0
+                df = pd.read_excel(file_path, sheet_name=sheet_name, header=start_row, dtype=str)
+
+                if header_row_idx is None and len(df) < 2:
+                    diagnostics.append(f"Sheet {sheet_name}: empty or unstructured, skipped")
+                    continue
+
+                # Clean keys.
+                df.columns = [str(c).strip() for c in df.columns]
+
+                # Iterate rows.
+                for idx, row in df.iterrows():
+                    row_data = {
+                        k: v
+                        for k, v in row.to_dict().items()
+                        if pd.notna(v) and str(v) != "nan" and v != ""
+                    }
+
+                    if not row_data:
+                        continue
+
+                    records.append({
+                        "type": "register_row",
+                        "data": {
+                            "sheet_name": sheet_name,
+                            "row_index": idx + start_row + 1,
+                            "content": row_data
+                        },
+                        "provenance": {"sheet": sheet_name, "row": idx + start_row + 1}
+                    })
+
+            diagnostics.append(f"Total register rows extracted: {len(records)}")
+            return ExtractionResult(records=records, diagnostics=diagnostics, success=True)
+
+        except Exception as e:
+            logger.error(f"Excel Extraction failed: {e}")
+            import traceback
+            diagnostics.append(str(e))
+            diagnostics.append(traceback.format_exc())
+            return ExtractionResult(success=False, diagnostics=diagnostics)

--- a/src/tests/test_excel_register_extractor_hygiene.py
+++ b/src/tests/test_excel_register_extractor_hygiene.py
@@ -1,0 +1,12 @@
+# -*- coding: utf-8 -*-
+from pathlib import Path
+
+
+def test_excel_register_extractor_does_not_write_absolute_debug_file():
+    source = Path("src/engine/extractors/register_extractor.py").read_text(encoding="utf-8-sig")
+
+    assert "D:\\SerapeumAI\\extractor_debug.txt" not in source
+    assert "extractor_debug.txt" not in source
+    assert "with open(" not in source
+    assert "diagnostics.append" in source
+    assert "logger.debug" in source

--- a/src/tests/test_extractor_registry_reachability.py
+++ b/src/tests/test_extractor_registry_reachability.py
@@ -1,0 +1,17 @@
+# -*- coding: utf-8 -*-
+
+from src.application.jobs.extract_job import ExtractJob
+from src.engine.extractors.dgn_extractor import DGNExtractor
+from src.engine.extractors.pptx_extractor import PPTXExtractor
+from src.engine.extractors.word_extractor import WordExtractor
+
+
+def test_extract_job_registry_includes_existing_reachable_document_extractors():
+    assert ExtractJob.EXTRACTORS["word"] is WordExtractor
+    assert ExtractJob.EXTRACTORS["pptx"] is PPTXExtractor
+    assert ExtractJob.EXTRACTORS["dgn"] is DGNExtractor
+
+
+def test_extract_job_registry_keeps_existing_evidence_extractors():
+    expected = {"p6", "ifc", "excel_register", "pdf", "field", "word", "pptx", "dgn"}
+    assert expected.issubset(set(ExtractJob.EXTRACTORS))

--- a/src/tests/test_ifc_extractor_persistence_contract.py
+++ b/src/tests/test_ifc_extractor_persistence_contract.py
@@ -1,0 +1,89 @@
+﻿# -*- coding: utf-8 -*-
+from pathlib import Path
+
+from src.application.jobs.extract_job import ExtractJob
+from src.infra.persistence.database_manager import DatabaseManager
+
+
+def _build_db(tmp_path):
+    db = DatabaseManager(root_dir=str(tmp_path), db_name=":memory:")
+    migration = Path("src/infra/persistence/migrations/001_baseline_v14.sql")
+    db.execute_script(migration.read_text(encoding="utf-8-sig"))
+
+    now = db._ts()
+    source_path = str(tmp_path / "model.ifc")
+
+    db.execute(
+        "INSERT INTO file_registry (file_id, project_id, first_seen_path, created_at) VALUES (?, ?, ?, ?)",
+        ("file_ifc", "ProjectA", source_path, now),
+    )
+    db.execute(
+        "INSERT INTO file_versions (file_version_id, file_id, sha256, size_bytes, file_ext, imported_at, source_path) VALUES (?, ?, ?, ?, ?, ?, ?)",
+        ("fv_ifc", "file_ifc", "hash-ifc", 1, ".ifc", now, source_path),
+    )
+    db.commit()
+
+    return db
+
+
+def test_ifc_element_metadata_records_are_persisted_to_ifc_elements(tmp_path):
+    db = _build_db(tmp_path)
+    job = ExtractJob("job_ifc", "ProjectA", "fv_ifc", extractor_name="ifc")
+
+    job._insert_record(
+        db,
+        {
+            "type": "ifc_element_metadata",
+            "data": {
+                "EntityType": "IfcWall",
+                "ElementId": "wall-guid-001",
+                "PSetName": "Pset_WallCommon",
+                "Properties": {"FireRating": "2HR"},
+                "IsQuantity": False,
+            },
+            "provenance": {"entity": "IfcWall", "pset": "Pset_WallCommon"},
+        },
+        doc_id="doc_ifc",
+    )
+
+    row = db.execute(
+        "SELECT element_id, entity_type, raw_properties_json FROM ifc_elements WHERE element_id=?",
+        ("wall-guid-001",),
+    ).fetchone()
+
+    assert row is not None
+    assert row["entity_type"] == "IfcWall"
+    assert "Pset_WallCommon" in row["raw_properties_json"]
+    assert "FireRating" in row["raw_properties_json"]
+
+
+def test_ifc_connection_records_are_persisted_to_links(tmp_path):
+    db = _build_db(tmp_path)
+    job = ExtractJob("job_ifc", "ProjectA", "fv_ifc", extractor_name="ifc")
+
+    job._insert_record(
+        db,
+        {
+            "type": "ifc_connection",
+            "data": {
+                "RelType": "Connectivity",
+                "Element1Id": "wall-guid-001",
+                "Element2Id": "slab-guid-002",
+            },
+            "provenance": {"entity": "IfcRelConnectsElements"},
+        },
+        doc_id="doc_ifc",
+    )
+
+    row = db.execute(
+        "SELECT project_id, link_type, from_kind, from_id, to_kind, to_id, status FROM links WHERE link_type=?",
+        ("ifc.connection",),
+    ).fetchone()
+
+    assert row is not None
+    assert row["project_id"] == "ProjectA"
+    assert row["from_kind"] == "ifc_element"
+    assert row["from_id"] == "wall-guid-001"
+    assert row["to_kind"] == "ifc_element"
+    assert row["to_id"] == "slab-guid-002"
+    assert row["status"] == "CANDIDATE"

